File: LeetCode/0117. Populating Next Right Pointers in Each Node II/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

"""
55 / 55 test cases passed.
Runtime: 56 ms
Memory Usage: 16.3 MB
"""
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root: None
        
        que = collections.deque([root])
        while que:
            prev = None
            for _ in range(len(que)):
                curr = que.popleft()
                if curr and curr.left: que.append(curr.left)
                if curr and curr.right: que.append(curr.right)
                if prev: prev.next = curr
                prev = curr
        return root

File: LeetCode/0117. Populating Next Right Pointers in Each Node II/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

/**
 * 55 / 55 test cases passed.
 * Runtime: 16 ms
 * Memory Usage: 17.2 MB 
 */
class Solution {
public:
    Node* connect(Node* root) {
        if (!root) return nullptr;
        
        queue<Node*> que;
        que.emplace(root);
        while (!que.empty()) {
            int layer = que.size();
            Node* prev = nullptr;
            for (int i = 0; i < layer; ++ i) {
                auto curr = que.front(); que.pop();
                if (curr->left) que.emplace(curr->left);
                if (curr->right) que.emplace(curr->right);
                if (i != 0) prev->next = curr;
                prev = curr;
            }
        }
        return root;
    }
};
